unit4/finalExercise.py

```
@@ -0,0 +1,190 @@
+        yield sec
+        sec += 1
+
+
+def gen_minutes() -> Generator:
+    """
+    Generates minutes in an hour
+    :return: a generator containing the minute
+    """
+    mnts: int = 0
+    while mnts < MIN_IN_HOUR:
+        yield mnts
+        mnts += 1
+
+
+def gen_hours() -> Generator:
+    """
+    Generates hours in a day
+    :return: a generator containing the hour
+    """
+    hour: int = 0
+    while hour < HOUR_IN_DAY:
+        yield hour
+        hour += 1
+
+
+def gen_time() -> Generator:
+    """
+    Generates the time in a day
+    :return: a generator containing a string that represents the time
+    """
+    end_of_clock: bool = False
+    sec_gen: Generator[int] = gen_secs()
+    mnts_gen: Generator[int] = gen_minutes()
+    hours_gen: Generator[int] = gen_hours()
+
+    sec: int = sec_gen.__next__()
+    mnts: int = mnts_gen.__next__()
+    hour: int = hours_gen.__next__()
+
+    yield f"{hour}:{mnts}:{sec}"
+    while not end_of_clock:
+        try:
+            sec = sec_gen.__next__()
+        except StopIteration:
+            try:
+                sec_gen = gen_secs()
+                sec = sec_gen.__next__()
+                mnts = mnts_gen.__next__()
+            except StopIteration:
+                mnts_gen = gen_minutes()
+                mnts = mnts_gen.__next__()
+                hour = hours_gen.__next__()
+
+        finally:
+            yield f"{hour}:{mnts}:{sec}"
+
+
+def gen_years(start: int = 2019) -> Generator:
+    """
+    Generates a year
+    :param start: the year the generator starts from
+    :return: a generator containing the year
+    """
+    year: int = start
+    while True:
+        yield year
+        year += 1
+
+
+def gen_months() -> Generator:
+    """
+    Generates month in a year
+    :return: a generator containing the month
+    """
+    month: int = 0
+    while month < MONTHS_IN_YEAR:
+        month += 1
+        yield month
+
+
+def gen_days(month: int, leap_year: bool = True) -> Generator:
```

I believe `leap_year` should default to `False` as the current value is quite misleading

---

This function's signature was given in the exercise (with the default values), should I still change it?

---

😢
Ok leave it as is

---

this is not perfect but does not matter much